model-serving-puller/puller/puller.go

```diff
@@ -79,73 +84,92 @@ func NewPullerFromConfig(log logr.Logger, config *PullerConfiguration) *Puller {
+		storageType, ok := modelKey.StorageParams[parameterKeyType]
+		if !ok {
+			return nil, fmt.Errorf("Predictor Storage field missing")
+		}
+		// the value for the default key should be populated from secrets associated with the Controller's ServiceAccount
+		defaultStorageKey := fmt.Sprintf("_%s_serviceaccount_secrets", storageType)
+
+		var err error
+		storageConfig, err = s.PullerConfig.GetStorageConfiguration(defaultStorageKey, s.Log)
+		if err != nil {
+			// do not error here, try to load from the parameters only
+			storageConfig = map[string]interface{}{}
+		}
+	} else {
+		var err error
+		storageConfig, err = s.PullerConfig.GetStorageConfiguration(*modelKey.StorageKey, s.Log)
+		if err != nil {
+			// if key was specified, but we could not find it, that is an error
+			return nil, fmt.Errorf("Did not find storage config for key %s: %w", *modelKey.StorageKey, err)
 		}
-	}
-	storageKey, ok := modelKey[jsonAttrModelKeyStorageKey].(string)
-	if !ok {
-		return nil, fmt.Errorf("Predictor Storage field missing")
 	}
 
-	storageConfig, err := s.PullerConfig.GetStorageConfiguration(storageKey, s.Log)
-	if err != nil {
-		return nil, err
+	// DEPRECATED: allow top-level bucket key for backwards compatibility
+	if modelKey.Bucket != "" {
+		storageConfig["bucket"] = modelKey.Bucket
 	}
 
-	// override storage config with per-request parameters
-	//  check "storage_params" first, but also check the top-level "bucket" key if
-	//  "storage_params" doesn't exist for backwards compatibility
-	if storageParams, ok := modelKey[jsonAttrStorageParams].(map[string]interface{}); ok {
-		if bucketName, ok1 := storageParams[jsonAttrModelKeyBucket]; ok1 {
-			storageConfig.Set("bucket", bucketName)
-		}
-	} else if bucketName, ok := modelKey[jsonAttrModelKeyBucket]; ok {
-		// return error if key exists but it is not a string
-		if _, ok1 := bucketName.(string); !ok1 {
-			return nil, fmt.Errorf("Invalid modelKey in LoadModelRequest, '%s' attribute must have a string value. Found value %v", jsonAttrModelKeyBucket, modelKey[jsonAttrModelKeyBucket])
-		} else {
-			storageConfig.Set("bucket", bucketName)
-		}
+	// override the storage configs with per-request storage parameters
+	if err := ApplyParameterOverrides(storageConfig, modelKey.StorageParams); err != nil {
+		return nil, fmt.Errorf("Unable to merge storage parameters from the storage config and the Predictor Storage field: %w", err)
 	}
 
-	modelDir, joinErr := util.SecureJoin(s.PullerConfig.RootModelDir, req.ModelId)
-	if joinErr != nil {
-		return nil, fmt.Errorf("Error joining paths '%s' and '%s': %v", s.PullerConfig.RootModelDir, req.ModelId, joinErr)
+	var storageType string
+	// if we still don't know the storage type, we cannot download the model, so return an error
+	storageType, ok := storageConfig[parameterKeyType].(string)
+	if !ok {
+		// return nil, fmt.Errorf("Unable to determine storage type. Either StorageKey must be set or StorageParameters must contain the 'type' parameter")
```

Comment:
I think this error would be more descriptive, but it breaks FVTs that check the error message.

Reply:
It just leads to a coupled change between the repos. I try to avoid those.
Is the coupled change the approach you want to take?

Reply:
Or we could change the FVT to permit either message?